mlir/lib/Dialect/Rock/utility/tosaUtils.cpp

```
@@ -0,0 +1,125 @@
+
+namespace mlir { 
+namespace rock {
+    
+bool isSpecificValueAttribute(Attribute value, double target) {
+  if (auto intValue = dyn_cast<IntegerAttr>(value))
+    return target == 0.0 ? intValue.getValue().isZero()
+                         : intValue.getValue() == target;
+  if (auto fpValue = dyn_cast<FloatAttr>(value))
+    return fpValue.getValue().isExactlyValue(target);
+  if (auto splatValue = dyn_cast<SplatElementsAttr>(value))
+    return isSpecificValueAttribute(splatValue.getSplatValue<Attribute>(),
+                                    target);
+  if (auto elementsValue = dyn_cast<ElementsAttr>(value))
+    return llvm::all_of(elementsValue.getValues<Attribute>(),
+                        [target](Attribute attr) {
+                          return isSpecificValueAttribute(attr, target);
+                        });
+  if (auto elementsValue = dyn_cast<DenseElementsAttr>(value))
+    return llvm::all_of(elementsValue.getValues<Attribute>(),
+                        [target](Attribute attr) {
+                          return isSpecificValueAttribute(attr, target);
+                        });
+  if (auto arrayValue = dyn_cast<ArrayAttr>(value))
+    return llvm::all_of(arrayValue.getValue(), [target](Attribute attr) {
+      return isSpecificValueAttribute(attr, target);
+    });
+  return false;
+}
+
+bool isConstantValue(Value v, double target) {
+  if (auto cst = v.getDefiningOp<arith::ConstantOp>())
+    return isSpecificValueAttribute(cst.getValue(), target);
+  if (auto cst = v.getDefiningOp<tosa::ConstOp>())
+    return isSpecificValueAttribute(cst.getValuesAttr(), target);
+  return false;
+}
+
+bool isConstantZero(Value v) {
+  auto elementTy = getElementTypeOrSelf(cast<ShapedType>(v.getType()));
+  if (isa<Float8E8M0FNUType>(elementTy)) {
+    // zero is not representable in Float8E8M0FNUType
+    LLVM_DEBUG(
+        llvm::dbgs()
+        << "Encountered Float8E8M0FNUType, which cannot represent zero.\n");
+    return false;
```

can we fail here instead of returning false?